File: ios/HybridImageView.swift
import Foundation
import UIKit
import Kingfisher
import NitroModules

/// Keeps processed (downsampled) images out of the disk cache: nothing to encode, nothing to write.
private struct MemoryOnlySerializer: CacheSerializer {
  func data(with image: KFCrossPlatformImage, original: Data?) -> Data? { nil }
  func image(with data: Data, options: KingfisherParsedOptionsInfo) -> KFCrossPlatformImage? { nil }
}

final class HybridImageView: HybridObitrainImageViewSpec, RecyclableView {
  let view: UIImageView = {
    let v = UIImageView()
    v.clipsToBounds = true
    v.isUserInteractionEnabled = false
    v.contentMode = .scaleAspectFill
    return v
  }()

  var uri: String? = nil
  var resource: String? = nil
  var resizeMode: NativeResizeMode? = nil
  var tintColor: Double? = nil
  var decodeWidth: Double? = nil
  var decodeHeight: Double? = nil
  var recyclingKey: String? = nil
  var onError: ((String) -> Void)? = nil

  private var loadedKey: String? = nil
  private var lastRecyclingKey: String? = nil


  // Kingfisher's `kf` API is @MainActor; Nitro may set props off-main.
  private func onMain(_ body: @escaping @MainActor () -> Void) {
    if Thread.isMainThread { MainActor.assumeIsolated { body() } } else { Task { @MainActor in body() } }
  }

  // Props arrive as one batch; load once per batch.
  func afterUpdate() {
    onMain { self.apply() }
  }

  @MainActor private func apply() {
    view.contentMode = contentMode(for: resizeMode ?? .cover)
    if recyclingKey != lastRecyclingKey {
      lastRecyclingKey = recyclingKey
      view.image = nil
      loadedKey = nil
    }
    let w = Int(decodeWidth ?? 0), h = Int(decodeHeight ?? 0)
    guard w > 0, h > 0 else { return } // not laid out yet
    let key = "\(uri ?? "")|\(resource ?? "")|\(w)x\(h)"
    if key != loadedKey {
      loadedKey = key
      load(w: w, h: h)
    }
    applyTint()
  }

  @MainActor private func load(w: Int, h: Int) {
    view.kf.cancelDownloadTask()
    if let name = resource, !name.isEmpty {
      view.image = UIImage(named: name)
      if view.image == nil { onError?("Bundled image not found: \(name)") }
      return
    }
    guard let s = uri, let url = URL(string: s) else { view.image = nil; return }
    // Kingfisher decodes to kCGImageSourceThumbnailMaxPixelSize = max(size) * scaleFactor.
    // JS sends pixels, so scaleFactor(1). cover gets headroom for non-matching aspect ratios.
    let headroom = (resizeMode ?? .cover) == .cover ? 1.25 : 1.0 // cover: shorter side must still fill the view
    let maxPx = CGFloat(Double(max(w, h)) * headroom)
    // ImageIO thumbnails are already decoded bitmaps, so no .backgroundDecode (it would redraw them).
    // The downsampled result stays in the memory cache only (serializer returns no data → no JPEG re-encode,
    // no disk write); the original bytes go to disk via .cacheOriginalImage so a new size never re-downloads.
    // (.cacheMemoryOnly + .cacheOriginalImage trips a Kingfisher fatalError in CacheCallbackCoordinator.)
    let options: KingfisherOptionsInfo = [
      .processor(DownsamplingImageProcessor(size: CGSize(width: maxPx, height: maxPx))),
      .scaleFactor(1),
      .cacheOriginalImage,
      .cacheSerializer(MemoryOnlySerializer()),
      .keepCurrentImageWhileLoading, // one layer commit per load instead of two (nil placeholder, then image)
    ]
    let source: Source = url.isFileURL
      ? .provider(LocalFileImageDataProvider(fileURL: url))
      : .network(url)
    view.kf.setImage(with: source, placeholder: nil, options: options) { [weak self] result in
      guard let self else { return }
      switch result {
      case .success(let r):
        self.applyTint()
      case .failure(let e):
        if !e.isTaskCancelled { self.onError?(e.localizedDescription) }
      }
    }
  }

  @MainActor private func applyTint() {
    guard let image = view.image else { return }
    if let t = tintColor {
      view.tintColor = Self.color(argb: t)
      if image.renderingMode != .alwaysTemplate { view.image = image.withRenderingMode(.alwaysTemplate) }
    } else if image.renderingMode == .alwaysTemplate {
      view.image = image.withRenderingMode(.alwaysOriginal)
    }
  }

  private func contentMode(for mode: NativeResizeMode) -> UIView.ContentMode {
    switch mode {
    case .cover: return .scaleAspectFill
    case .contain: return .scaleAspectFit
    case .stretch: return .scaleToFill
    case .center: return .center
    }
  }

  private static func color(argb: Double) -> UIColor {
    let v = UInt32(truncatingIfNeeded: Int64(argb))
    return UIColor(
      red: CGFloat((v >> 16) & 0xff) / 255,
      green: CGFloat((v >> 8) & 0xff) / 255,
      blue: CGFloat(v & 0xff) / 255,
      alpha: CGFloat((v >> 24) & 0xff) / 255)
  }

  func onDropView() {
    onMain { self.view.kf.cancelDownloadTask() }
  }

  // Reset everything: props are only re-applied on a recycled view if the next consumer sets them.
  func prepareForRecycle() {
    onMain { self.view.kf.cancelDownloadTask() }
    view.image = nil
    uri = nil; resource = nil; resizeMode = nil; tintColor = nil
    decodeWidth = nil; decodeHeight = nil; recyclingKey = nil; onError = nil
    loadedKey = nil; lastRecyclingKey = nil
  }
}

File: android/src/main/java/com/margelo/nitro/obitrain/reactnativeimage/HybridImageView.kt
package com.margelo.nitro.obitrain.reactnativeimage

import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.widget.ImageView
import androidx.appcompat.widget.AppCompatImageView
import coil3.request.ImageRequest
import coil3.request.allowHardware
import coil3.request.target
import coil3.size.Precision
import coil3.util.CoilUtils
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.views.RecyclableView

@DoNotStrip
class HybridImageView(val context: ThemedReactContext) : HybridObitrainImageViewSpec(), RecyclableView {
  // The request target must be the ImageView itself so Coil sizes with Precision.INEXACT.
  override val view: AppCompatImageView = AppCompatImageView(context).apply {
    scaleType = ImageView.ScaleType.CENTER_CROP
    isClickable = false
    isFocusable = false
  }

  override var uri: String? = null
  override var resource: String? = null
  override var resizeMode: NativeResizeMode? = null
  override var tintColor: Double? = null
  override var decodeWidth: Double? = null
  override var decodeHeight: Double? = null
  override var recyclingKey: String? = null
  override var onError: ((String) -> Unit)? = null

  private var loadedKey: String? = null
  private var lastRecyclingKey: String? = null

  // Props arrive as one batch; load once per batch.
  override fun afterUpdate() {
    view.scaleType = scaleType(resizeMode ?: NativeResizeMode.COVER)
    view.colorFilter = tintColor?.let { PorterDuffColorFilter(it.toLong().toInt(), PorterDuff.Mode.SRC_IN) }
    if (recyclingKey != lastRecyclingKey) {
      lastRecyclingKey = recyclingKey
      view.setImageDrawable(null)
      loadedKey = null
    }
    val w = decodeWidth?.toInt() ?: 0
    val h = decodeHeight?.toInt() ?: 0
    if (w <= 0 || h <= 0) return // not laid out yet
    val key = "$uri|$resource|${w}x$h"
    if (key == loadedKey) return
    loadedKey = key
    load(w, h)
  }

  private fun load(w: Int, h: Int) {
    val data: Any = resource?.takeIf { it.isNotEmpty() }?.let { name ->
      val id = context.resources.getIdentifier(name, "drawable", context.packageName)
      if (id == 0) { onError?.invoke("Bundled image not found: $name"); return }
      id
    } ?: uri ?: run { view.setImageDrawable(null); return }
    val request = ImageRequest.Builder(context)
      .data(data)
      .target(view)
      .size(w, h)
      .precision(Precision.INEXACT)
      .allowHardware(true)
      .listener(onError = { _, r -> onError?.invoke(r.throwable.message ?: "Image load failed") })
      .build()
    ImageLoaders.get(context).enqueue(request) // cancels the previous request on this view
  }

  private fun scaleType(mode: NativeResizeMode): ImageView.ScaleType = when (mode) {
    NativeResizeMode.COVER -> ImageView.ScaleType.CENTER_CROP
    NativeResizeMode.CONTAIN -> ImageView.ScaleType.FIT_CENTER
    NativeResizeMode.STRETCH -> ImageView.ScaleType.FIT_XY
    NativeResizeMode.CENTER -> ImageView.ScaleType.CENTER
  }

  override fun onDropView() {
    CoilUtils.dispose(view)
  }

  // Reset everything: props are only re-applied on a recycled view if the next consumer sets them.
  override fun prepareForRecycle() {
    CoilUtils.dispose(view)
    view.setImageDrawable(null)
    view.colorFilter = null
    uri = null; resource = null; resizeMode = null; tintColor = null
    decodeWidth = null; decodeHeight = null; recyclingKey = null; onError = null
    loadedKey = null; lastRecyclingKey = null
  }
}
